File: WalletKitCore/src/crypto/BRCryptoWalletSweeper.c
//
//  BRCryptoWalletSweeper.c
//  BRCore
//
//  Created by Ehsan Rezaie on 5/22/20
//  Copyright © 2020 breadwallet. All rights reserved.
//
//  See the LICENSE file at the project root for license information.
//  See the CONTRIBUTORS file at the project root for a list of contributors.

#include "BRCryptoWalletP.h"
#include "BRCryptoWalletSweeperP.h"
#include "BRCryptoHandlersP.h"


extern BRCryptoWalletSweeper
cryptoWalletSweeperAllocAndInit (size_t sizeInBytes,
                                 BRCryptoBlockChainType type,
                                 BRCryptoKey key,
                                 BRCryptoUnit unit) {
    assert (sizeInBytes >= sizeof (struct BRCryptoWalletSweeperRecord));
    assert (cryptoKeyHasSecret (key));
    
    BRCryptoWalletSweeper sweeper = calloc (1, sizeInBytes);
    
    sweeper->type = type;
    sweeper->handlers = cryptoHandlersLookup(type)->sweeper;
    sweeper->key = cryptoKeyTake (key);
    sweeper->unit = cryptoUnitTake (unit);
    
    return sweeper;
}

extern void
cryptoWalletSweeperRelease (BRCryptoWalletSweeper sweeper) {
    
    sweeper->handlers->release (sweeper);
    
    cryptoKeyGive (sweeper->key);
    cryptoUnitGive (sweeper->unit);
    
    memset (sweeper, 0, sizeof(*sweeper));
    free (sweeper);
}

extern BRCryptoWalletSweeperStatus
cryptoWalletSweeperAddTransactionFromBundle (BRCryptoWalletSweeper sweeper,
                                             OwnershipKept BRCryptoClientTransactionBundle bundle) {
    if (NULL == sweeper->handlers->addTranactionFromBundle) {
        assert(0);
        return CRYPTO_WALLET_SWEEPER_ILLEGAL_OPERATION;
    }
    
    return sweeper->handlers->addTranactionFromBundle (sweeper, bundle);
}

extern void
cryptoWalletManagerEstimateFeeBasisForWalletSweep (BRCryptoWalletSweeper sweeper,
                                                   BRCryptoWalletManager cwm,
                                                   BRCryptoWallet wallet,
                                                   BRCryptoCookie cookie,
                                                   BRCryptoNetworkFee fee) {
    BRCryptoFeeBasis feeBasis = sweeper->handlers->estimateFeeBasis (cwm,
                                                                     wallet,
                                                                     cookie,
                                                                     sweeper,
                                                                     fee);
    
    if (NULL != feeBasis)
        cryptoWalletGenerateEvent (wallet, cryptoWalletEventCreateFeeBasisEstimated (CRYPTO_SUCCESS, cookie, feeBasis));

    cryptoFeeBasisGive(feeBasis);
}

extern BRCryptoTransfer
cryptoWalletSweeperCreateTransferForWalletSweep (BRCryptoWalletSweeper sweeper,
                                                 BRCryptoWalletManager cwm,
                                                 BRCryptoWallet wallet,
                                                 BRCryptoFeeBasis estimatedFeeBasis) {
    BRCryptoTransfer transfer = sweeper->handlers->createTransfer (cwm, wallet, sweeper, estimatedFeeBasis);
    
    cryptoTransferGenerateEvent (transfer, (BRCryptoTransferEvent) {
        CRYPTO_TRANSFER_EVENT_CREATED
    });
    
    return transfer;
}

extern BRCryptoKey
cryptoWalletSweeperGetKey (BRCryptoWalletSweeper sweeper) {
    return cryptoKeyTake (sweeper->key);
}

extern BRCryptoAddress
cryptoWalletSweeperGetAddress (BRCryptoWalletSweeper sweeper) {
    return sweeper->handlers->getAddress (sweeper);
}

extern BRCryptoAmount
cryptoWalletSweeperGetBalance (BRCryptoWalletSweeper sweeper) {
    return sweeper->handlers->getBalance (sweeper);
}

extern BRCryptoWalletSweeperStatus
cryptoWalletSweeperValidate (BRCryptoWalletSweeper sweeper) {
    return sweeper->handlers->validate (sweeper);
}

File: WalletKitJava/CoreNative/src/main/java/com/breadwallet/corenative/crypto/BRCryptoWalletSweeper.java
/*
 * Created by Michael Carrara <michael.carrara@breadwallet.com> on 7/1/19.
 * Copyright (c) 2019 Breadwinner AG.  All right reserved.
*
 * See the LICENSE file at the project root for license information.
 * See the CONTRIBUTORS file at the project root for a list of contributors.
 */
package com.breadwallet.corenative.crypto;

import com.breadwallet.corenative.CryptoLibraryDirect;
import com.breadwallet.corenative.utility.SizeT;
import com.google.common.base.Optional;
import com.sun.jna.Pointer;
import com.sun.jna.PointerType;

public class BRCryptoWalletSweeper extends PointerType {

    public BRCryptoWalletSweeper() {
        super();
    }

    public BRCryptoWalletSweeper(Pointer address) {
        super(address);
    }

    public static BRCryptoWalletSweeperStatus validateSupported(BRCryptoWalletManager cwm,
                                                                BRCryptoWallet wallet,
                                                                BRCryptoKey key) {
        return BRCryptoWalletSweeperStatus.fromCore(
                CryptoLibraryDirect.cryptoWalletManagerWalletSweeperValidateSupported(
                        cwm.getPointer(),
                        wallet.getPointer(),
                        key.getPointer()
                )
        );
    }

    public static BRCryptoWalletSweeper createAsBtc(BRCryptoWalletManager cwm,
                                                    BRCryptoWallet wallet,
                                                    BRCryptoKey key) {
        return new BRCryptoWalletSweeper(
                CryptoLibraryDirect.cryptoWalletManagerCreateWalletSweeper(
                        cwm.getPointer(),
                        wallet.getPointer(),
                        key.getPointer()
                )
        );
    }

    public BRCryptoKey getKey() {
        Pointer thisPtr = this.getPointer();

        return new BRCryptoKey(CryptoLibraryDirect.cryptoWalletSweeperGetKey(thisPtr));
    }

    public Optional<BRCryptoAmount> getBalance() {
        Pointer thisPtr = this.getPointer();

        return Optional.fromNullable(
                CryptoLibraryDirect.cryptoWalletSweeperGetBalance(
                        thisPtr
                )
        ).transform(BRCryptoAmount::new);
    }

    public Optional<BRCryptoAddress> getAddress() {
        Pointer thisPtr = this.getPointer();

        return Optional.fromNullable(
                CryptoLibraryDirect.cryptoWalletSweeperGetAddress(thisPtr)
        ).transform(BRCryptoAddress::new);
    }

    public BRCryptoWalletSweeperStatus handleTransactionAsBtc(byte[] transaction) {
        Pointer thisPtr = this.getPointer();

        //TODO:SWEEP use transaction bundle
        return BRCryptoWalletSweeperStatus.fromCore(
                CryptoLibraryDirect.cryptoWalletSweeperAddTransactionFromBundle(
                        thisPtr,
                        transaction,
                        new SizeT(transaction.length))
        );
    }

    public BRCryptoWalletSweeperStatus validate() {
        Pointer thisPtr = this.getPointer();

        return BRCryptoWalletSweeperStatus.fromCore(
                CryptoLibraryDirect.cryptoWalletSweeperValidate(thisPtr)
        );
    }

    public void give() {
        Pointer thisPtr = this.getPointer();

        CryptoLibraryDirect.cryptoWalletSweeperRelease(thisPtr);
    }
}
